File: Sources/ClipboardCore/ClipStore.swift
import Foundation

public final class ClipStore {
    public let storageDirectory: URL
    public private(set) var items: [ClipItem]

    private let now: () -> Date
    private let fileManager: FileManager
    private let metadataURL: URL
    private let imageDirectory: URL
    private let expirationInterval: TimeInterval

    public init(
        storageDirectory: URL,
        now: @escaping () -> Date = Date.init,
        fileManager: FileManager = .default,
        expirationInterval: TimeInterval = 7 * 24 * 60 * 60
    ) throws {
        self.storageDirectory = storageDirectory
        self.now = now
        self.fileManager = fileManager
        self.expirationInterval = expirationInterval
        self.metadataURL = storageDirectory.appendingPathComponent("clips.json")
        self.imageDirectory = storageDirectory.appendingPathComponent("Images", isDirectory: true)

        try fileManager.createDirectory(at: storageDirectory, withIntermediateDirectories: true)
        try fileManager.createDirectory(at: imageDirectory, withIntermediateDirectories: true)

        if fileManager.fileExists(atPath: metadataURL.path) {
            let data = try Data(contentsOf: metadataURL)
            self.items = try JSONDecoder.clipStore.decode([ClipItem].self, from: data)
        } else {
            self.items = []
        }
    }

    @discardableResult
    public func addText(_ text: String) throws -> ClipItem {
        let trimmedPreview = text
            .replacingOccurrences(of: "\n", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let preview = trimmedPreview.isEmpty ? "Empty text" : trimmedPreview
        let clip = ClipItem(
            type: .text,
            createdAt: now(),
            text: text,
            previewText: preview,
            contentHash: ContentHasher.textHash(text)
        )

        return try upsert(clip)
    }

    @discardableResult
    public func addImageData(
        _ data: Data,
        fileExtension: String,
        previewText: String = "Image"
    ) throws -> ClipItem {
        let hash = ContentHasher.imageHash(data)
        if let existing = items.first(where: { $0.contentHash == hash }) {
            return try refreshExistingClip(existing)
        }

        let normalizedExtension = fileExtension.trimmingCharacters(in: .whitespacesAndNewlines)
        let safeExtension = normalizedExtension.isEmpty ? "png" : normalizedExtension
        let relativePath = "Images/\(UUID().uuidString).\(safeExtension)"
        let imageURL = storageDirectory.appendingPathComponent(relativePath)
        try data.write(to: imageURL, options: .atomic)

        let clip = ClipItem(
            type: .image,
            createdAt: now(),
            imagePath: relativePath,
            previewText: previewText,
            contentHash: hash
        )

        return try upsert(clip)
    }

    public func setPinned(_ id: UUID, _ isPinned: Bool) throws {
        guard let index = items.firstIndex(where: { $0.id == id }) else {
            return
        }

        items[index].isPinned = isPinned
        try save()
    }

    public func markCopied(_ id: UUID) throws {
        guard let index = items.firstIndex(where: { $0.id == id }) else {
            return
        }

        items[index].lastCopiedAt = now()
        let clip = items.remove(at: index)
        items.insert(clip, at: 0)
        try save()
    }

    public func delete(_ id: UUID) throws {
        guard let index = items.firstIndex(where: { $0.id == id }) else {
            return
        }

        let clip = items.remove(at: index)
        try deleteImageFileIfNeeded(for: clip)
        try save()
    }

    public func cleanupExpiredClips() throws {
        let cutoff = now().addingTimeInterval(-expirationInterval)
        let expired = items.filter { !$0.isPinned && $0.createdAt < cutoff }
        items.removeAll { !$0.isPinned && $0.createdAt < cutoff }

        for clip in expired {
            try deleteImageFileIfNeeded(for: clip)
        }

        try save()
    }

    public func clips(matching query: String, filter: ClipFilter = .all) -> [ClipItem] {
        let normalizedQuery = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return items.filter { clip in
            let matchesFilter: Bool = switch filter {
            case .all:
                true
            case .text:
                clip.type == .text
            case .image:
                clip.type == .image
            case .pinned:
                clip.isPinned
            }
            let matchesQuery = normalizedQuery.isEmpty
                || clip.previewText.lowercased().contains(normalizedQuery)
                || (clip.text?.lowercased().contains(normalizedQuery) ?? false)
            return matchesFilter && matchesQuery
        }
    }

    private func upsert(_ clip: ClipItem) throws -> ClipItem {
        if let existing = items.first(where: { $0.contentHash == clip.contentHash }) {
            return try refreshExistingClip(existing)
        }

        items.insert(clip, at: 0)
        try save()
        return clip
    }

    private func refreshExistingClip(_ existing: ClipItem) throws -> ClipItem {
        guard let index = items.firstIndex(where: { $0.id == existing.id }) else {
            return existing
        }

        var refreshed = items.remove(at: index)
        refreshed.createdAt = now()
        items.insert(refreshed, at: 0)
        try save()
        return refreshed
    }

    private func save() throws {
        let data = try JSONEncoder.clipStore.encode(items)
        try data.write(to: metadataURL, options: .atomic)
    }

    private func deleteImageFileIfNeeded(for clip: ClipItem) throws {
        guard let imagePath = clip.imagePath else {
            return
        }

        let url = storageDirectory.appendingPathComponent(imagePath)
        if fileManager.fileExists(atPath: url.path) {
            try fileManager.removeItem(at: url)
        }
    }
}

private extension JSONEncoder {
    static var clipStore: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }
}

private extension JSONDecoder {
    static var clipStore: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}

File: ClipboardSS.Windows/src/ClipboardSS.Core/Storage/ClipStore.cs
using System.Text.Json;
using ClipboardSS.Core.Crypto;
using ClipboardSS.Core.Json;
using ClipboardSS.Core.Models;

namespace ClipboardSS.Core.Storage;

public sealed class ClipStore
{
    public static readonly TimeSpan DefaultExpiration = TimeSpan.FromDays(7);

    private readonly object _gate = new();
    private readonly Func<DateTimeOffset> _now;
    private readonly TimeSpan _expiration;
    private readonly string _metadataPath;
    private readonly string _imageDirectory;
    private List<ClipItem> _items;

    public ClipStore(
        string storageDirectory,
        Func<DateTimeOffset>? now = null,
        TimeSpan? expiration = null)
    {
        StorageDirectory = Path.GetFullPath(storageDirectory);
        _now = now ?? (() => DateTimeOffset.UtcNow);
        _expiration = expiration ?? DefaultExpiration;
        _metadataPath = Path.Combine(StorageDirectory, "clips.json");
        _imageDirectory = Path.Combine(StorageDirectory, "Images");
        Directory.CreateDirectory(StorageDirectory);
        Directory.CreateDirectory(_imageDirectory);
        _items = File.Exists(_metadataPath)
            ? JsonSerializer.Deserialize<List<ClipItem>>(File.ReadAllBytes(_metadataPath), WireJson.Options)
                ?? []
            : [];
    }

    public string StorageDirectory { get; }

    public IReadOnlyList<ClipItem> Items
    {
        get
        {
            lock (_gate)
            {
                return _items.ToArray();
            }
        }
    }

    public ClipItem AddText(string text)
    {
        var trimmedPreview = text.Replace("\n", " ", StringComparison.Ordinal).Trim();
        var clip = new ClipItem
        {
            Type = ClipType.Text,
            CreatedAt = _now(),
            Text = text,
            PreviewText = trimmedPreview.Length == 0 ? "Empty text" : trimmedPreview,
            ContentHash = ContentHasher.TextHash(text),
        };

        lock (_gate)
        {
            return Upsert(clip);
        }
    }

    public ClipItem AddImageData(
        ReadOnlySpan<byte> data,
        string fileExtension,
        string previewText = "Image")
    {
        var hash = ContentHasher.ImageHash(data);
        lock (_gate)
        {
            var existing = _items.FirstOrDefault(item => item.ContentHash == hash);
            if (existing is not null)
            {
                return RefreshExisting(existing);
            }

            var extension = fileExtension.Trim();
            if (extension.Length == 0)
            {
                extension = "png";
            }

            var fileName = $"{Guid.NewGuid():D}.{extension}";
            var relativePath = Path.Combine("Images", fileName);
            AtomicFile.WriteAllBytes(Path.Combine(StorageDirectory, relativePath), data);
            var clip = new ClipItem
            {
                Type = ClipType.Image,
                CreatedAt = _now(),
                ImagePath = relativePath,
                PreviewText = previewText,
                ContentHash = hash,
            };
            return Upsert(clip);
        }
    }

    public void SetPinned(Guid id, bool isPinned)
    {
        lock (_gate)
        {
            var clip = _items.FirstOrDefault(item => item.Id == id);
            if (clip is null)
            {
                return;
            }

            clip.IsPinned = isPinned;
            Save();
        }
    }

    public void MarkCopied(Guid id)
    {
        lock (_gate)
        {
            var index = _items.FindIndex(item => item.Id == id);
            if (index < 0)
            {
                return;
            }

            var clip = _items[index];
            clip.LastCopiedAt = _now();
            _items.RemoveAt(index);
            _items.Insert(0, clip);
            Save();
        }
    }

    public void Delete(Guid id)
    {
        lock (_gate)
        {
            var index = _items.FindIndex(item => item.Id == id);
            if (index < 0)
            {
                return;
            }

            var clip = _items[index];
            _items.RemoveAt(index);
            DeleteImageFileIfNeeded(clip);
            Save();
        }
    }

    public void CleanupExpiredClips()
    {
        lock (_gate)
        {
            var cutoff = _now() - _expiration;
            var expired = _items.Where(item => !item.IsPinned && item.CreatedAt < cutoff).ToArray();
            if (expired.Length == 0)
            {
                return;
            }
            _items.RemoveAll(item => !item.IsPinned && item.CreatedAt < cutoff);
            foreach (var clip in expired)
            {
                DeleteImageFileIfNeeded(clip);
            }

            Save();
        }
    }

    public IReadOnlyList<ClipItem> Clips(string query, ClipFilter filter = ClipFilter.All)
    {
        var normalizedQuery = query.Trim().ToLowerInvariant();
        lock (_gate)
        {
            return _items.Where(clip =>
            {
                var matchesFilter = filter switch
                {
                    ClipFilter.All => true,
                    ClipFilter.Text => clip.Type == ClipType.Text,
                    ClipFilter.Image => clip.Type == ClipType.Image,
                    ClipFilter.Pinned => clip.IsPinned,
                    _ => false,
                };
                var matchesQuery = normalizedQuery.Length == 0
                    || clip.PreviewText.Contains(normalizedQuery, StringComparison.OrdinalIgnoreCase)
                    || (clip.Text?.Contains(normalizedQuery, StringComparison.OrdinalIgnoreCase) ?? false);
                return matchesFilter && matchesQuery;
            }).ToArray();
        }
    }

    private ClipItem Upsert(ClipItem clip)
    {
        var existing = _items.FirstOrDefault(item => item.ContentHash == clip.ContentHash);
        if (existing is not null)
        {
            return RefreshExisting(existing);
        }

        _items.Insert(0, clip);
        Save();
        return clip;
    }

    private ClipItem RefreshExisting(ClipItem existing)
    {
        var index = _items.FindIndex(item => item.Id == existing.Id);
        if (index < 0)
        {
            return existing;
        }

        _items.RemoveAt(index);
        existing.CreatedAt = _now();
        _items.Insert(0, existing);
        Save();
        return existing;
    }

    private void Save() =>
        AtomicFile.WriteAllBytes(
            _metadataPath,
            JsonSerializer.SerializeToUtf8Bytes(_items, WireJson.IndentedOptions));

    private void DeleteImageFileIfNeeded(ClipItem clip)
    {
        if (clip.ImagePath is null)
        {
            return;
        }

        var path = Path.Combine(StorageDirectory, clip.ImagePath);
        if (File.Exists(path))
        {
            File.Delete(path);
        }
    }
}
